File: js/typed/Step.js

class StepLeft {

  constructor() {
    this.action = null;
    Object.seal(this);
  }

  doMove(call = true) {
    this.action.next.cell = this.action.current.cell.left();
    if (call) this.action.doNext();
    else return [this.action, 'doNext'];
  }

  toString() {
    return "LEFT";
  }
}

class StepRight {

  constructor() {
    this.action = null;
    Object.seal(this);
  }

  doMove(call = true) {
    this.action.next.cell = this.action.current.cell.right();
    if (call) this.action.doNext();
    else return [this.action, 'doNext'];
  }

  toString() {
    return "RIGHT";
  }
}

class StepNone {

  constructor() {
    this.action = null;
    Object.seal(this);
  }

  doMove(call = true) {
    this.action.next.cell = this.action.current.cell;
    if (call) this.action.doNext();
    else return [this.action, 'doNext'];
  }

  toString() {
    return "NONE";
  }
}

module.exports = {StepLeft, StepRight, StepNone};

File: c/typed/step.c

#include "Step.h"

#include "action.h"
#include "State.h"
#include "Cell.h"

void StepLeft::do_move() {
  action->next->cell = action->current->cell->left();
  action->do_next();
}

void StepLeft::print(std::ostream& os) const {
  os << "LEFT";
}

void StepRight::do_move() {
  action->next->cell = action->current->cell->right();
  action->do_next();
}

void StepRight::print(std::ostream& os) const {
  os << "RIGHT";
}

void StepNone::do_move() {
  action->next->cell = action->current->cell;
  action->do_next();
}

void StepNone::print(std::ostream& os) const {
  os << "NONE";
}

std::ostream& operator<<(std::ostream& os, const Step* step) {
  step->print(os);
  return os;
}
